### PR TITLE
polish print_tree

### PR DESCRIPTION
**Pull request by DKormann**
shows Lazybuffer in tree. usefull for understanding codegen.
can now print any lazyop or buffer not just tensor
no new lines

### CODE REVIEW

**Comment by wozeparrot** [comment]
Can you send an example of a before and after?

File: extra/utils.py [hunk 1]
```diff
@@ -209,12 +209,12 @@ def get_child(parent, key):
...
+  if type(lazydata).__name__ == "LazyBuffer": return [f"━━ realized {lazydata.dtype.name} {lazydata.shape}"] if (lazydata.realized) else _tree(lazydata.op, "LB ")
+  if len(lazydata.src) == 0: return [f"━━ {prefix}{lazydata.op.name} {lazydata.arg if lazydata.arg else ''}"]
+  lines = [f"━┳ {prefix}{lazydata.op.name} {lazydata.arg if lazydata.arg else ''}"]
   childs = [_tree(c) for c in lazydata.src[:]]
   for c in childs[:-1]: lines += [f" ┣{c[0]}"] + [f" ┃{l}" for l in c[1:]]
   return lines + [" ┗"+childs[-1][0]] + ["  "+l for l in childs[-1][1:]]
 
-def print_tree(tensor:Tensor):print("\n".join([f"{str(i).rjust(3)} {s}" for i,s in enumerate(_tree(tensor.lazydata))]))
\ No newline at end of file
+def print_tree(tensor:Tensor):print("\n".join([f"{str(i).rjust(3)} {s}" for i,s in enumerate(_tree(tensor if type(tensor) != Tensor else tensor.lazydata))]))
```
Comment (wozeparrot) [inline extra/utils.py:220]: this should be `isinstance` not a comparison against `type()`

Comment (DKormann) [reply extra/utils.py]:
```
a = Tensor.cat(*[Tensor([1]) for _ in range(4)])
print_tree(a)
```

Comment (DKormann) [reply extra/utils.py]:
```
  0 ━┳ LB ADD 
  1  ┣━┳ ADD 
  2  ┃ ┣━┳ ADD 
  3  ┃ ┃ ┣━┳ LB PAD ((0, 3),)
  4  ┃ ┃ ┃ ┗━━ realized float (1,)
  5  ┃ ┃ ┗━┳ LB PAD ((1, 2),)
  6  ┃ ┃   ┗━━ realized float (1,)
  7  ┃ ┗━┳ LB PAD ((2, 1),)
  8  ┃   ┗━━ realized float (1,)
  9  ┗━┳ LB PAD ((3, 0),)
 10    ┗━━ realized float (1,)
```

Comment (DKormann) [reply extra/utils.py]: before it just doesnt have any 'LB'

File: extra/utils.py [hunk 2]
```diff
@@ -209,12 +209,12 @@ def get_child(parent, key):
       obj = getattr(obj, k)
   return obj
 
-def _tree(lazydata):
-  if type(lazydata).__name__ == "LazyBuffer": return [f"━━ realized {lazydata.dtype.name} {lazydata.shape}"] if (lazydata.realized) else _tree(lazydata.op)
-  if len(lazydata.src) == 0: return [f"━━ {lazydata.op.name} {lazydata.arg if lazydata.arg else ''}"]
-  lines = [f"━┳ {lazydata.op.name} {lazydata.arg if lazydata.arg else ''}"]
+def _tree(lazydata, prefix=""):
+  if type(lazydata).__name__ == "LazyBuffer": return [f"━━ realized {lazydata.dtype.name} {lazydata.shape}"] if (lazydata.realized) else _tree(lazydata.op, "LB ")
```
Comment (roelofvandijk) [inline extra/utils.py:213]: `lazydata,__class__ is LazyBuffer`

File: extra/utils.py [hunk 3]
```diff
@@ -209,12 +209,12 @@ def get_child(parent, key):
...
+  if len(lazydata.src) == 0: return [f"━━ {prefix}{lazydata.op.name} {lazydata.arg if lazydata.arg else ''}"]
+  lines = [f"━┳ {prefix}{lazydata.op.name} {lazydata.arg if lazydata.arg else ''}"]
   childs = [_tree(c) for c in lazydata.src[:]]
   for c in childs[:-1]: lines += [f" ┣{c[0]}"] + [f" ┃{l}" for l in c[1:]]
   return lines + [" ┗"+childs[-1][0]] + ["  "+l for l in childs[-1][1:]]
 
-def print_tree(tensor:Tensor):print("\n".join([f"{str(i).rjust(3)} {s}" for i,s in enumerate(_tree(tensor.lazydata))]))
\ No newline at end of file
+def print_tree(tensor:Tensor):print("\n".join([f"{str(i).rjust(3)} {s}" for i,s in enumerate(_tree(tensor if not isinstance(tensor, Tensor) else tensor.lazydata))]))
```
Comment (roelofvandijk) [inline extra/utils.py:220]: this could maybe be `tensor.getattr('lazydata', tensor)`

**Pull request merged**